config/initializers/flipper.rb

```diff
@@ -0,0 +1,16 @@
+require "flipper"
+require "flipper/adapters/active_record"
+
+Flipper.configure do |config|
+  config.default do
+    Flipper.new(Flipper::Adapters::ActiveRecord.new)
+  end
+end
+Rails.configuration.middleware.use Flipper::Middleware::Memoizer, preload_all: true
+
+Flipper.register(:admins) { |actor| actor.respond_to?(:admin?) && actor.admin? }
+
+if !Flipper[:unit_price].exist?
```

Comment:
I looked into the backtrace and this is the line that fails when trying to run the migration. So I guess that the lesson here is that we can't access or initialise data in an initializer.

I think that we have to move this part somewhere else. Maybe we shouldn't alter the `feature?` helper but modify our `FeatureToggle` class to set defaults and then call Flipper. Keeping FeatureToggle as a wrapper for Flipper also has the advantage that we only need to change that one class if we want to replace Flipper with something else later.

Reply:
> I looked into the backtrace and this is the line that fails when trying to run the migration. So I guess that the lesson here is that we can't access or initialise data in an initializer.

Yes, that's something I noticed. A little bit disappointed, but, yeah, it's not possible _by design_.

Reply:
Actually, on this part, I'm a bit incompetent and I should request some help.

Reply:
I checked it and migrations fail because Flipper requires them to work. I removed these lines, migrated, restored them, and I can see unit prices locally again. However, why do we exactly need this defensive code? does the line below conflict with anything you set through Flipper UI? is that it?

Reply:
Btw, you replace `if !whatever` with `unless whatever` 👌

Reply:
> I checked it and migrations fail because Flipper requires them to work. I removed these lines, migrated, restored them, and I can see unit prices locally again.

Yes, that's the point.

> However, why do we exactly need this defensive code? does the line below conflict with anything you set through Flipper UI? is that it?

 - We can indeed define the feature `:unit_price` within the UI, but I wanted to initialize feature. 
 - But at least we need to define what is the admins group, and it's not possible with the UI....

Reply:
Hey @jibees, I hope you don't mind me adding some commits to your branch. I added my idea how we can have backwards compatibility and how we don't need to access the database in the initializer. What do you think?

Feel free to disregard my changes or squash some commits (especially the reverted one). It's just adding ideas to your work in progress. It was easier for me to show code instead of explaining it in text. I'm also happy to have a call at some point (just no time today).

Reply:
Hey @mkllnk thanks a lot for this: I really appreciate you can put some help here. Feel free to add as many commits you wants on my branches ;)


Here's my comments: 

 - Conditional testing inside `lib/open_food_network/feature_toggle.rb` seems to be relevant and I think it's a good idea!
 - We don't initialize a `unit_prices` feature, but maybe it's not that important (and instance manager can handle that inside the Flipper interface)
 - I'm more annoyed with this block, in `config/initializers/feature_toggles.rb` : 

```ruby
OpenFoodNetwork::FeatureToggle.enable(:unit_price) do |user|
  ['development', 'staging'].include?(ENV['RAILS_ENV']) || user.try(:admin?)
end
```
Here, we _initialize_ a feature (`unit_price`) which is totally unlinked with Flipper, and can't be handled by it. 
So, I'm wondering what is the purpose of this one?
Actually, if I had a feature with the same name inside the FlipperUI it's handled by Flipper itself. But it can be very confusing at least.

 - I've deleted commit fdc67dc57e89e3dd876ce814c0c78cf9c1478433 and fdc67dc57e89e3dd876ce814c0c78cf9c1478433 as they cancelled each other out 
- Migration was ok on my side, but not on the CI. I need to investigate this one.

Reply:
> we initialize a feature (unit_price) which is totally unlinked with Flipper, and can't be handled by it.
So, I'm wondering what is the purpose of this one?

I know it's not linked to Flipper but I thought that it's a very efficient way of providing a default when Flipper doesn't know about that feature yet. But I guess that you would like the feature to appear in the Flipper UI automatically, right? We could do this in `config/initializers/feature_toggles.rb`:
```rb
OpenFoodNetwork::FeatureToggle.enable(:unit_price) do |user|
  Flipper.enable_group(:unit_price, :admins)
  Flipper.enabled?(:unit_price, user)
end
```

Reply:
The main default of this one (correct me if I'm wrong) is that `unit_price` feature is not initialized until this block is called ie. the first time we call `OpenFoodNetwork::FeatureToggle.enabled?`. Thus, flipper is initialized and the we directly call `Flipper.enabled?(feature_name, user)` in `lib/open_food_network/feature_toggle.rb`. 

I'm wondering if we shouldn't remove this part of code, and let's instance managers initialized itself each feature. I can see few arguments in favor of doing this: 
 - Feature toggle configuration is fully managed in the UI, and we avoid several entry points
 - Instance manager decide on when/how they enable each feature
 - A feature can't be enabled if instance manager is not aware of this one: he must know about it

Reply:
> Feature toggle configuration is fully managed in the UI, and we avoid several entry points
> Instance manager decide on when/how they enable each feature

Yes, that's a good approach. I like that to reduce confusion and give instances the most freedom.

> A feature can't be enabled if instance manager is not aware of this one: he must know about it.

Yes, so we need a way to add features without flipping them. We could add new features in migrations:
```rb
class AddFeatureUnitPrice < ActiveRecord::Migration[5.0]
  class FlipperFeature < ActiveRecord::Base; end

  def up
    FlipperFeature.create!(key: "unit_price")
  end

  def down
    FlipperFeature.delete_all(key: "unit_price")
  end
end
```

Reply:
@mkllnk I've push the last version without any initialization. Thanks a lot! 🙏